src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,442 @@
+    }
+  }
+}
+
+// Get history weights
+void ArpaSampling::ComputeHistoriesWeights() {
+  for (auto it = histories_.begin(); it != histories_.end(); ++it) {
+    HistType history(*(it));
+    KALDI_ASSERT(history.size() <= ngram_order_);
+    for (int32 i = 0; i < history.size() + 1; i++) {
+      HistType h_tmp = history;
+      BaseFloat prob = 1.0 / histories_.size();
+      while (h_tmp.size() > (history.size() - i)) {
+        HistType::iterator last = h_tmp.end() - 1;
+        HistType h(h_tmp.begin(), last);
+        int32 word = h_tmp.back();
+        prob *= pow(10, GetBackoffWeight(h_tmp.size(), word, h));
+        HistType h_up(h_tmp.begin() + 1, h_tmp.end());
+        h_tmp = h_up;
+      }
+      HistType::iterator begin = history.begin() + i;
+      HistType h(begin, history.end());
+      hists_weights_[h] += prob;
+    }
+  }
+  KALDI_LOG << "Size of hists_weights_ is: " << hists_weights_.size();
+}
+
+// Get weighted pdf
+void ArpaSampling::ComputeWeightedPdf(std::vector<std::pair<int32, BaseFloat> >* pdf_w) {
+  BaseFloat prob = 0;
+  (*pdf_w).resize(num_words_); // if do not do this, (*pdf_w)[word] += prob will get seg fault
+  for (int32 i = 0; i < num_words_; i++) {
+    for (auto it = hists_weights_.begin(); it != hists_weights_.end(); ++it) {
+      HistType h(it->first);
+      int32 order = h.size();
+      auto it_hist = probs_[order].find(h);
+      if (it_hist != probs_[order].end()) {
+        int32 word = vocab_[i].second;
+        auto it_word = probs_[order][h].find(word);
+        if (it_word != probs_[order][h].end()) {
+          if (order > 0) {
+            HistType::iterator last = h.end() - 1;
+            HistType::iterator first = h.begin() + 1;
+            HistType h1(h.begin(), last);
+            HistType h2(first, h.end());
+            prob = (it->second) * (pow(10, probs_[order][h][word].first) - 
+                    pow(10, GetBackoffWeight(order, h.back(), h1))
+                    * pow(10, GetProb(order, word, h2)));
+            (*pdf_w)[i].first = word;
+            (*pdf_w)[i].second += prob;
+          } else {
+            prob = (it->second) * pow(10, probs_[order][h][word].first);
+            (*pdf_w)[i].first = word;
+            (*pdf_w)[i].second += prob;
+          }
+        }
+      }
+    } // end reading history
+  } // end reading words
+}
+
+// sample a word that follows a pdf
+int32 ArpaSampling::SampleWord(const std::vector<std::pair<int32, BaseFloat> >& pdf) {
+  // generate a cdf from the given pdf 
+  std::vector<std::pair<int32, BaseFloat> > cdf;
+  BaseFloat upper = 0;
+  int32 word;
+  std::pair<int32, BaseFloat> probs;
+  for (int32 i = 0; i < num_words_; i++) {
+    upper += pdf[i].second;
+    word = vocab_[i].second;
+    probs = std::make_pair(word, upper);
+    cdf.push_back(probs);
+  }
+  BaseFloat u = 1.0 * RandUniform();
+  if (u >= 0 && u < cdf[1].second) {
+    return cdf[0].first;
+  }
+  for (int32 i = 1; i < num_words_; i++) {
```

Comment:
change this to a binary search. Check my SelectOne() function in rnnlm-utils.cc for reference if you need.